src/RelayClient.ts

```diff
@@ -0,0 +1,56 @@
+import { BigNumber, BigNumberish, getDefaultProvider, providers } from 'ethers';
+import { BigNumber as BigNumberJs } from 'bignumber.js';
+import type { EnvelopingTypes } from '@rsksmart/rif-relay-contracts';
+import { ESTIMATED_GAS_CORRECTION_FACTOR } from './constants/relay.const';
+
+class RelayClient {
+    private readonly _provider: providers.Provider;
+
+    constructor() {
+        this._provider = getDefaultProvider();
+    }
+
+    async estimateInternalCallGas(
+        { request, relayData }: EnvelopingTypes.RelayRequestStruct,
+        internalGasCorrection: BigNumberish,
+        addCorrection = true
+    ): Promise<BigNumber> {
+
+        let estimation = await this._provider.estimateGas({
+            from: request.from,
+            to: request.to,
+            gasPrice: relayData.gasPrice,
+            data: request.data,
+        });
+
+        // The INTERNAL_TRANSACTION_ESTIMATE_CORRECTION is substracted because the estimation is done using web3.eth.estimateGas which
+        // estimates the call as if it where an external call, and in our case it will be called internally (it's not the same cost).
+        // Because of this, the estimated maxPossibleGas in the server (which estimates the whole transaction) might not be enough to successfully pass
+        // the following verification made in the SmartWallet:
+        // require(gasleft() > req.gas, "Not enough gas left"). This is done right before calling the destination internally
+        if (estimation.gt(internalGasCorrection)) {
+            estimation = estimation.sub(internalGasCorrection);
+        }
+
+        if (addCorrection) {
+            estimation = this._applyGasCorrectionFactor(estimation);
+        }
+
+        return estimation;
+    }
+
+    private _applyGasCorrectionFactor(estimation: BigNumberish): BigNumber {
+        let bigEstimation = BigNumberJs(estimation.toString());
+        const bigGasCorrection = BigNumberJs(
+            ESTIMATED_GAS_CORRECTION_FACTOR
+        );
+
+        if (!bigGasCorrection.isEqualTo(1)) {
```

Comment:
What's the harm in letting it apply a correction factor of 1?

Reply:
also this check is unnecessary unless ESTIMATED_GAS_CORRECTION_FACTOR is a configuration property.

Reply:
ESTIMATED_GAS_CORRECTION_FACTOR, its a configuration property.